src/Sarif.Viewer.VisualStudio/Tags/SarifLocationTextMarkerTagger.cs

```diff
@@ -0,0 +1,196 @@
+        public IEnumerable<ITagSpan<ITextMarkerTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (this.tagsDirty)
+            {
+                this.tagsDirty = false;
+
+                this.UnsubscribeFromTagEvents();
+
+                SarifErrorListItem currentlySelectedItem = this.sarifErrorListEventSelectionService.SelectedItem;
+
+                // We need to make sure the list isn't modified underneath us while providing the tags, so executing ToList to get our copy.
+                this.currentTags = (currentlySelectedItem == null ? Enumerable.Empty<ISarifLocationTag>() :
+                    CodeAnalysisResultManager.
+                    Instance.
+                    RunIndexToRunDataCache.
+                    Values.
+                    SelectMany(runDataCache => runDataCache.SarifErrors).
+                    Where(sarifListItem => sarifListItem.ResultId == currentlySelectedItem.ResultId).
+                    Where(sarifListItem => string.Compare(this.filePath, sarifListItem.FileName, StringComparison.OrdinalIgnoreCase) == 0).
+                    SelectMany(sarifListItem => sarifListItem.GetTags<ITextMarkerTag>(this.textBuffer, this.persistentSpanFactory, includeChildTags: true, includeResultTag: true))).
+                    ToList();
+
+                // We need to subscribe to property change events from the provided tags
+                // because they will change their highlight color depending on whether
+                // the editor caret is within them.
+                // When they change their colors, we need to ask Visual Studio to refresh it's tags.
+                this.SubscribeToTagEvents();
+            }
+
+            if (!this.currentTags.Any())
+            {
+                yield break;
+            }
+
+            foreach (SnapshotSpan span in spans)
+            {
+                foreach (var possibleTag in this.currentTags.Where(possibleTag => possibleTag.DocumentPersistentSpan.Span != null))
+                {
+                    SnapshotSpan possibleTagSnapshotSpan = possibleTag.DocumentPersistentSpan.Span.GetSpan(span.Snapshot);
+                    if (span.IntersectsWith(possibleTagSnapshotSpan))
+                    {
+                        yield return new TagSpan<ITextMarkerTag>(possibleTagSnapshotSpan, (ITextMarkerTag)possibleTag);
+                    }
+                }
+            }
+        }
+
+        public void RefreshTags()
+        {
+            this.tagsDirty = true;
+
+            ITextSnapshot textSnapshot = this.textBuffer.CurrentSnapshot;
+            this.TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(new SnapshotSpan(textSnapshot, 0, textSnapshot.Length)));
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void SubscribeToTagEvents()
+        {
+            if (this.currentTags == null || !this.currentTags.Any())
```

Comment:
>(this.currentTags == null || !this.currentTags.Any() [](start = 15, length = 52)

Use ?.Any() == false

Reply:
this.currentTags?.Any()  != true

---
In reply to: [502098816](https://github.com/microsoft/sarif-visualstudio-extension/pull/214#discussion_r502098816) [](ancestors = 502098816)